File: src/main/java/com/nerdery/jvm/resistance/bots/hopper/HopperBayesianBot.java
package com.nerdery.jvm.resistance.bots.hopper;

import com.nerdery.jvm.resistance.bots.DoctorBot;
import com.nerdery.jvm.resistance.models.Prescription;
import twitter4j.*;
import twitter4j.auth.AccessToken;
import twitter4j.auth.Authorization;

import java.io.IOException;
import java.io.InputStream;
import java.text.Format;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * The Hopper bot. Uses conditional probabilities.
 *
 * Pros: Tries to use probabilities of outcomes when making decisions.
 * Cons: Other doctors are illogical, and sometimes this bot is just unlucky.
 *
 * @author Stephen Hopper
 */
public class HopperBayesianBot implements DoctorBot {

    private static final int MIN_ROUNDS_FOR_DOC_PROB_CALC = 9;
    private static final float DEFAULT_DOC_PROB_CALC = 0.9f;
    private static final int MIN_SIMILAR_PRESCRIPTION_PROB_CALC = 5;
    private static final float WEIGHT_SIMILAR_PRESCRIPTION_PROB = 0.5f;
    private static final float WEIGHT_OVERALL_PRESCRIPTION_PROB = (1 - WEIGHT_SIMILAR_PRESCRIPTION_PROB);

    private static final int MONEY_ANTIBIOTIC_GOOD = 3;
    private static final int MONEY_ANTIBIOTIC_BAD = -100;
    private static final int MONEY_REST_GOOD = 1;
    private static final int MONEY_REST_BAD = -10;

    private static final Twitter twitter;

    static {
        twitter = TwitterFactory.getSingleton();
    }

    private Map<String, List<PrescriptionData>> prescriptionMap = new HashMap<>();

    @Override
    public String getUserId() {
        return "hopper";
    }

    @Override
    public boolean prescribeAntibiotic(float patientTemperature, Collection<Prescription> previousPrescriptions) {
        updatePrescriptions(previousPrescriptions);
        PatientTemp temp = PatientTemp.lookupTemp(patientTemperature);
        float probMoneyIfPrescribeAntibiotic = probMoneyGivenAntibiotic(temp);
        float probMoneyIfPrescribeRest = probMoneyGivenRest(temp);

        System.out.println("Patient temp: " + patientTemperature);
        System.out.println("Probability of money if prescribe antibiotic: " + probMoneyIfPrescribeAntibiotic);
        System.out.println("Probability of money if prescribe rest: " + probMoneyIfPrescribeRest);

//        float prescribeAntibioticMoneyFactor = probMoneyIfPrescribeAntibiotic * (MONEY_ANTIBIOTIC_BAD + MONEY_ANTIBIOTIC_GOOD);
//        float prescribeRestMoneyFactor = probMoneyIfPrescribeRest * (MONEY_REST_BAD + MONEY_REST_GOOD);
//        System.out.println("Prescribe antibiotic money factor: " + prescribeAntibioticMoneyFactor);
//        System.out.println("Prescribe rest money factor: " + prescribeRestMoneyFactor);
//
//        boolean prescribeDrugs = prescribeAntibioticMoneyFactor >= prescribeRestMoneyFactor;
        boolean prescribeDrugs = probMoneyIfPrescribeAntibiotic > probMoneyIfPrescribeRest;
        System.out.println("Prescribe antibiotics?: " + prescribeDrugs);

        StatusUpdate status = buildStatus(patientTemperature, prescribeDrugs, probMoneyIfPrescribeAntibiotic, probMoneyIfPrescribeRest);
        try {
            twitter.updateStatus(status);
        } catch (TwitterException e) {
            e.printStackTrace();
        }

        return prescribeDrugs;
    }

    private StatusUpdate buildStatus(float temp, boolean prescribeDrugs, float chanceMoneyAnti, float chanceMoneyRest) {
        String status;
        String antiMoneyStr = format(chanceMoneyAnti * 100.0f);
        String restMoneyStr = format(chanceMoneyRest * 100.0f);
        if (prescribeDrugs) {
            status = "Prescribed drugs to a patient with a fever of " + format(temp) +
                    "F as there's a " + antiMoneyStr + "% chance I get $$$ from this (vs " + restMoneyStr + "%) #NerderyResistance";
        } else {
            status = "Told some fool to rest up because their fever was " + format(temp) +
                    "F and there's a " + restMoneyStr + "% chance I get $$$ from this (vs " + antiMoneyStr + "%) #NerderyResistance";
        }
        System.out.println(status);

        StatusUpdate update = new StatusUpdate(status);
        update.setPossiblySensitive(false); //hilarious
        return update;
    }

    private String format(float f) {
        return String.format("%.1f", f);
    }

    private void updatePrescriptions(Collection<Prescription> previousPrescriptions) {
        previousPrescriptions.
                stream().
                filter(x -> !"hopper".equals(x.getUserId())).
                forEach(x -> {
                    if (prescriptionMap.containsKey(x.getUserId())) {
                        prescriptionMap.get(x.getUserId()).add(PrescriptionData.fromPrescription(x));
                    } else {
                        List<PrescriptionData> prescriptions = new ArrayList<>();
                        prescriptions.add(PrescriptionData.fromPrescription(x));
                        prescriptionMap.put(x.getUserId(), prescriptions);
                    }
                });
    }

    private float probMoneyGivenAntibiotic(PatientTemp temp) {
        return probInfected(temp) + (1.0f - probInfected(temp)) * probLuckyGivenAntibiotic(temp);
    }

    private float probLuckyGivenAntibiotic(PatientTemp temp) {
        return 1.0f - probUnluckyGivenAntibiotic(temp);
    }

    private float probUnluckyGivenAntibiotic(PatientTemp temp) {
        if (countRounds() <= MIN_ROUNDS_FOR_DOC_PROB_CALC) {
            return DEFAULT_DOC_PROB_CALC;
        } else {
            return prescriptionMap.
                    entrySet().
                    stream().
                    map(entries -> {
                        List<PrescriptionData> prescriptions = entries.getValue();
                        float overallProbPrescribe = probPrescribeAntibiotic(prescriptions);
                        Stream<PrescriptionData> similarPrescriptionData = prescriptions.
                                stream().
                                filter(x -> x.getTemp() == temp);
                        if (similarPrescriptionData.count() >= MIN_SIMILAR_PRESCRIPTION_PROB_CALC) {
                            float similarPrescriptionProb = probPrescribeAntibiotic(similarPrescriptionData.collect(Collectors.toList()));
                            return WEIGHT_SIMILAR_PRESCRIPTION_PROB * similarPrescriptionProb + WEIGHT_OVERALL_PRESCRIPTION_PROB * overallProbPrescribe;
                        } else {
                            return overallProbPrescribe;
                        }
                    }).reduce(1.0f, (x, y) -> x * y);
        }
    }

    private float probPrescribeAntibiotic(List<PrescriptionData> samples) {
        return samples.stream().filter(PrescriptionData::isPrescribedAntibiotics).count() / (float) samples.size();
    }

    private int countRounds() {
        return prescriptionMap.
                entrySet().
                stream().
                map(x -> x.getValue().size()).
                max(Integer::compareTo).
                orElse(0);
    }

    //rest probability methods
    private float probMoneyGivenRest(PatientTemp temp) {
        return probInfected(temp) * probLuckyGivenRest(temp) + (1.0f - probInfected(temp));
    }

    private float probLuckyGivenRest(PatientTemp temp) {
        return 1.0f - probUnluckyGivenRest(temp);
    }

    private float probUnluckyGivenRest(PatientTemp temp) {
        return probInfected(temp) * randy();
    }

    private float randy() {
        return new Random().nextFloat();
    }

    //general probability methods
    private float probInfected(PatientTemp temp) {
        return temp.getProbInfected();
    }
}

File: src/main/scala/com/nerdery/jvm/resistance/bots/entrants/hopper/HopperBayesianBot.scala
package com.nerdery.jvm.resistance.bots.entrants.hopper

import java.security.SecureRandom
import java.util

import akka.actor._
import akka.actor.{Actor, ActorLogging, ActorSystem, Props}
import com.nerdery.jvm.resistance.bots.DoctorBot
import com.nerdery.jvm.resistance.models.Prescription
import twitter4j.{StatusUpdate, TwitterFactory}

import scala.collection.JavaConversions._
import scala.collection.mutable
import scala.collection.mutable.ListBuffer

/**
  * The Hopper bot. Uses conditional probabilities. Also violates
  * HIPAA by Tweeting about patients its seeing.
  *
  * Pros: Tries to use probabilities of outcomes when making decisions. Is a social butterfly knife.
  * Cons: Other doctors are illogical, and sometimes this bot is just unlucky. But those things don't matter,
  * it was going to get sued anyways for violating HIPAA.
  *
  * Note: Twitter functionality assumes there's a twitter4j.properties file on the root of the classpath.
  *
  * @author Stephen Hopper (stephenmhopper@gmail.com)
  */
class HopperBayesianBot(tweetMuch: Boolean = true) extends DoctorBot {
  import HopperBayesianBot._
  private val prescriptionMap = mutable.Map[String, ListBuffer[PrescriptionData]]()
  private val randy = new SecureRandom()

  //Default constructor. Disables Twitter functionality.
  def this() {
    this(false)
  }

  override def getUserId: String = "hopper"

  override def prescribeAntibiotic(patientTemperature: Float, previousPrescriptions: util.Collection[Prescription]) = {
    updatePrescriptions(previousPrescriptions.toList)
    val temp: PatientTemp = PatientTemp(patientTemperature)
    val probMoneyIfPrescribeAntibiotic: Float = probMoneyGivenAntibiotic(temp)
    val probMoneyIfPrescribeRest: Float = probMoneyGivenRest(temp)

    println(s"Patient temp: $patientTemperature")
    println(s"Probability of money if prescribe antibiotic: $probMoneyIfPrescribeAntibiotic")
    println(s"Probability of money if prescribe rest: $probMoneyIfPrescribeRest")

    //do the comparison. If it's within 5%, then we'll just go with antibiotics 'cause we get moar $$$
    val prescribeDrugs = probMoneyIfPrescribeAntibiotic > (probMoneyIfPrescribeRest - 0.05f)
    println(s"Prescribe antibiotics?: $prescribeDrugs")

    val status: StatusUpdate = buildStatus(patientTemperature, prescribeDrugs, probMoneyIfPrescribeAntibiotic, probMoneyIfPrescribeRest)
    val otherDocStatus = buildOtherDocStatusMaybe(previousPrescriptions.toList)
    if (tweetMuch) {
      try {
        twitterActor ! status
        otherDocStatus.foreach(s => twitterActor ! s)
      } catch {
        case t: Throwable =>
          println(s"Something went wrong whilst tweeting: $t")
          t.printStackTrace()
      }
    }

    prescribeDrugs
  }

  private def buildOtherDocStatusMaybe(previousPrescriptions: List[Prescription]): Option[StatusUpdate] = {
    if (previousPrescriptions.nonEmpty) {
      val otherDocs = previousPrescriptions.filter(_.getUserId != "hopper")
      val otherDoc = otherDocs(randy.nextInt(otherDocs.size))
      val handle = s"@${otherDoc.getUserId.substring(0,Math.min(49, otherDoc.getUserId.length))}"
      val temp = PatientTemp(otherDoc.getTemperature)
      val tempStr = s"${format(temp.temperature)}F"
      val statusOpt: Option[String] = (PatientTemp(otherDoc.getTemperature), otherDoc.isPrescribedAntibiotics) match {
        case (t, true) if t.getProbInfected <= 0.25 => Option(s"OMG ROFL $handle prescribed drugs to a patient with a temp of $tempStr $HASHTAG")
        case (t, false) if t.getProbInfected <= 0.25 => Option(s"$handle prescribed rest to a patient with a temp of $tempStr. Such dull Much boring $HASHTAG")
        case (t, false) if t.getProbInfected > 0.5 => Option(s"Lolz $handle prescribed rest to a patient with a temp of $tempStr $HASHTAG")
        case _ => None
      }
      statusOpt.map { status =>
        println(status)
        val update = new StatusUpdate(status)
        update.setPossiblySensitive(false) //hilarious
        update
      }
    } else {
      None
    }
  }

  private def buildStatus(temp: Float, prescribeDrugs: Boolean, chanceMoneyAnti: Float, chanceMoneyRest: Float): StatusUpdate = {
    val antiMoneyStr: String = format(chanceMoneyAnti * 100.0f)
    val restMoneyStr: String = format(chanceMoneyRest * 100.0f)
    val status: String = prescribeDrugs match {
      case true => s"Prescribed drugs to a patient with a fever of ${format(temp)}F as there's a $antiMoneyStr% chance I get $$$$$$ from this (vs $restMoneyStr%) $HASHTAG"
      case false => s"Told some fool to rest up because their fever was ${format(temp)}F and there's a $restMoneyStr% chance I get $$$$$$ from this (vs $antiMoneyStr%) $HASHTAG"
    }
    println(status)
    val update: StatusUpdate = new StatusUpdate(status)
    update.setPossiblySensitive(false) //hilarious
    update
  }

  private def format(f: Float): String = "%.1f".format(f)

  private def updatePrescriptions(previousPrescriptions: List[Prescription]) {
    previousPrescriptions.filter(p => !"hopper".equals(p.getUserId)).foreach(p => {
      val prescriptionData = PrescriptionData.apply(p)
      prescriptionMap.get(p.getUserId) match {
        case Some(prescriptions) => prescriptions += prescriptionData
        case _ => {
          val prescriptions = ListBuffer[PrescriptionData](prescriptionData)
          prescriptionMap.put(p.getUserId, prescriptions)
        }
      }
    })
  }

  private def probMoneyGivenAntibiotic(temp: PatientTemp): Float = probInfected(temp) + (1.0f - probInfected(temp)) * probLuckyGivenAntibiotic(temp)

  private def probLuckyGivenAntibiotic(temp: PatientTemp): Float = 1.0f - probUnluckyGivenAntibiotic(temp)

  private def probUnluckyGivenAntibiotic(temp: PatientTemp): Float = {
    countRounds <= MIN_ROUNDS_FOR_DOC_PROB_CALC match {
      case true => DEFAULT_DOC_PROB_CALC
      case false => {
        val probs = prescriptionMap.map {
          case (docId, prescriptions) =>
            val overallProbPrescribe = probPrescribeAntibiotic(prescriptions.toList)
            val similarPrescriptionData = prescriptions.filter(_.temp == temp).toList
            similarPrescriptionData.size >= MIN_SIMILAR_PRESCRIPTION_PROB_CALC match {
              case true =>
                val similarPrescriptionProb = probPrescribeAntibiotic(similarPrescriptionData)
                WEIGHT_SIMILAR_PRESCRIPTION_PROB * similarPrescriptionProb + WEIGHT_OVERALL_PRESCRIPTION_PROB * overallProbPrescribe
              case false => overallProbPrescribe
            }
        }
        probs.foldLeft(1.0f)(_ * _)
      }
    }
  }

  private def probPrescribeAntibiotic(samples: List[PrescriptionData]): Float = samples.count(_.prescribedAntibiotics) / samples.size.toFloat

  private def countRounds: Int = prescriptionMap.map(entry => entry._2.size).reduceOption(_ max _).getOrElse(0)

  private def probMoneyGivenRest(temp: PatientTemp): Float = probInfected(temp) * probLuckyGivenRest(temp) + (1.0f - probInfected(temp))

  private def probLuckyGivenRest(temp: PatientTemp): Float = 1.0f - probUnluckyGivenRest(temp)

  private def probUnluckyGivenRest(temp: PatientTemp): Float = math.pow((temp.temperature - 100.0) / 3.0, 5.0).toFloat / 2.0f

  private def probInfected(temp: PatientTemp): Float = temp.getProbInfected
}

object HopperBayesianBot {
  private val MIN_ROUNDS_FOR_DOC_PROB_CALC: Int = 9
  private val DEFAULT_DOC_PROB_CALC: Float = 0.9f
  private val MIN_SIMILAR_PRESCRIPTION_PROB_CALC: Int = 5
  private val WEIGHT_SIMILAR_PRESCRIPTION_PROB: Float = 0.9f
  private val WEIGHT_OVERALL_PRESCRIPTION_PROB: Float = 1 - WEIGHT_SIMILAR_PRESCRIPTION_PROB
  private val HASHTAG = "#NerderyResistance"

  //sure, we could just use futures, but why not over-engineer and just use Akka?
  private lazy val system = ActorSystem("HopperBayes")
  private lazy val twitterActor = system.actorOf(Props(classOf[TweetMuch]), "tweet-much")

  class TweetMuch extends Actor with ActorLogging {
    def receive = {
      case status: StatusUpdate =>
        log.info("Sending status update to twitter")
        TwitterFactory.getSingleton.updateStatus(status)
    }
  }
}